lib/spack/spack/hooks/module_file_generation.py

```diff
@@ -11,24 +11,36 @@
 
 def _for_each_enabled(spec, method_name):
     """Calls a method for each enabled module"""
-    enabled = spack.config.get('modules:enable')
-    if not enabled:
-        tty.debug('NO MODULE WRITTEN: list of enabled module files is empty')
-        return
-
-    for name in enabled:
-        generator = spack.modules.module_types[name](spec)
-        try:
-            getattr(generator, method_name)()
-        except RuntimeError as e:
-            msg = 'cannot perform the requested {0} operation on module files'
-            msg += ' [{1}]'
-            tty.warn(msg.format(method_name, str(e)))
+    for name in spack.config.get('modules', {}):
+        enabled = spack.config.get('modules:%s:enable' % name)
+        if not enabled:
+            tty.debug('NO MODULE WRITTEN: list of enabled module files is empty')
+            return
+
+        for type in enabled:
+            generator = spack.modules.module_types[type](spec, name)
+            try:
+                getattr(generator, method_name)()
+            except RuntimeError as e:
+                msg = 'cannot perform the requested {0} operation on module files'
+                msg += ' [{1}]'
+                tty.warn(msg.format(method_name, str(e)))
 
 
 def post_install(spec):
-    _for_each_enabled(spec, 'write')
+    # If the spec is installed through an environment, we delegate to
+    # the post_env_write hook so that spack can manage interactions between
+    # env views and modules
+    import spack.environment  # break import cycle
+    env = spack.environment.get_env({}, '')
+    if not env:
+        _for_each_enabled(spec, 'write')
```

Comment:
Nitpick:

```suggestion
    if env:
        return  # skip and leave this to post_env_write
        
    _for_each_enabled(spec, 'write')
```

The reason I'd do it this way is b/c I tend to put the exceptions first and the meat of the function last, as I expect the latter to evolve. I don't think the `return` is likely to change, but the rest of the function seems more likely to, so I'd rather it not be nested.